collections/_article/d3d12-shaders-with-a-twist.md

```diff
@@ -0,0 +1,112 @@
+
+## Translation of GLSL/SPIR-V to DXIL
+
+Godot rendering architecture is built with the industry standard Vulkan graphics API in mind. And that is great. However, once you want to plug new rendering APIs into the engine, you'll find that some elements need to be wired in a manner that they work with the alternative API (Direct3D 12 in this case.)
+
+Shaders in Godot are written in its custom, very **GLSL**-like language. (The builtin shaders are already GLSL in the source code.) Shader code in this language is compiled by the engine (thanks to **glslang**) into SPIR-V. which is a well-defined format that stores the binary representation of a shader.
+
+SPIR-V and Vulkan play nice. However, Direct3D 12 features its own binary shader format: **DXIL**. This format is structured as a container with its most important contained chunks being shader metadata and the compiled shader code in the form of **LLVM bitcode** (yes, the same LLVM/Clang uses for an intermediate representation of the compiled code.) As a consequence, this rendering driver would need a way to generate DXIL from either the GLSL or the SPIR-V representations of the Godot shaders.
+
+### First round: SPIRV-Cross
+
+After assessing the available third-party libraries and open source projects capable of doing that (and license-compatible), I found **SPIRV-Cross** was the most sensible choice: it would translate GLSL to HLSL (the high level counterpart of GLSL in the DirectX world). From that point, the D3D12 renderer would use the standard **DirectX Shader Compiler** library to compile HLSL into DXIL.
+
+That worked quite well, but wasn't without big headaches about writing the code that should match the resource bindings in the Vulkan-targeted shaders to those in the DXIL ones. I won't delve into details on that here. Just know it required an insane amount of code and relying on certain SPIRV-Cross specifics. For now, the takeaway is that it was a hell of an approach, but worked in practice! (If you're curious, in the [first version of the pull request](https://github.com/godotengine/godot/pull/64304), you can see for yourself what I'm talking about.) This process also allowed me to make a few little contributions to SPIRV-Cross itself and learn its internals, which is always a good thing.
+
+### Second round: Via Mesa's NIR
+
+Months went by and I verified that the Microsoft was making great progress with **Dozen**, the Vulkan-to-D3D12 translation layer found in **Mesa**.
+
+Mesa was already capable of translating SPIR-V into its own intermediate representation (**NIR**), but Dozen would have added the ability to translate NIR to DXIL. That looked very promising because it would be a binary-to-binary translation (i.e., SPIR-V to DXIL) without having to go all the way back to a high level language (GLSL/HLSL), which is slower and less convenient.
```

Comment:
```suggestion
Mesa was already capable of translating SPIR-V into its own intermediate representation (**NIR**), but Dozen would have added the ability to translate NIR to DXIL. That looked very promising because it would be a binary-to-binary translation (i.e., SPIR-V to DXIL). This would avoid having to go all the way back to a high level-language (GLSL/HLSL), which is slower and less convenient.
```

Reply:
Applied, but I put the hyphen in "high-level," not in "level-language," as I think your intent was.